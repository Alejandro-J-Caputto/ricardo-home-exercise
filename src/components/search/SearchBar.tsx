import React from "react";
import { useEffect } from "react";
import { useState } from "react";
import { useHistory } from "react-router-dom";

const SearchBar: React.FC<{}> = (props) => {
  const history = useHistory();

  const [enteredText, setEnteredText] = useState("");

  const [searchInputIsValid, setSearchInputIsValid] = useState<boolean>(false);

  const inputChangeHandler = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEnteredText(event.target.value);
  };
  const submitHandler = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
    history.push(`/home-exercise/search/${enteredText}`);
  };

  useEffect(() => {
    const validTimer = setTimeout(() => {
      setSearchInputIsValid(enteredText.trim().length > 3);
    }, 500);
    return () => {
      clearInterval(validTimer);
    };
  }, [enteredText]);

  return (
    <div className="searchBar">
      <form className="searchBar-group">
        <input
          onChange={inputChangeHandler}
          type="text"
          name="search"
          id="search"
          placeholder="Search Article"
          className="searchBar-group__input"
          autoComplete="off"
        />
        <label htmlFor="search" className="searchBar-group__label">
          Search Article
        </label>
        <button
          onClick={submitHandler}
          className="btn btn-search"
          disabled={!searchInputIsValid}
        >
          <i className="fas fa-search"></i>
          <span>Search</span>
        </button>
      </form>
    </div>
  );
};
export default SearchBar;
