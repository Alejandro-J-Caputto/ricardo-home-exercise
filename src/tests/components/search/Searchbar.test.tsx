import { shallow, configure } from "enzyme";

import "@testing-library/jest-dom";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import SearchBar from "../../../components/search/SearchBar";
import { createSerializer } from "enzyme-to-json";
import React from "react";

jest.mock("react-router-dom", () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

configure({ adapter: new Adapter() });
//@ts-ignore
expect.addSnapshotSerializer(createSerializer({ mode: "deep", noKey: true }));
const eneteredText = "";
const searchInputIsValid = false;
React.useState = jest
  .fn()
  .mockReturnValueOnce([eneteredText, {}])
  .mockReturnValueOnce([searchInputIsValid, {}]);
const wrapper = shallow(<SearchBar />);
const searchInput = wrapper.find("input");
describe("Tests on searchBar component", () => {
  test("should render Searchbar correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Should have a label "search Article', () => {
    const label = wrapper.find("label").text();
    expect(label).toBe("Search Article");
  });

  test("The button should be initialy disabled", () => {
    const button = wrapper.find("button");
    expect(button.props().disabled).toBe(true);
  });

  test("should change input value", () => {
    const value = "hello Ricardo team";
    searchInput.simulate("change", { target: { value } });
    //CREATE AN ELEMENT INSIDE THE COMPONENT TO CHECK THIS ASSERTION
  });
  test("should submit the form", (done) => {
    const value = "hello Ricardo team";

    searchInput.simulate("change", { target: { value } });

    setTimeout(() => {
      wrapper.find("button").simulate("click", { preventDefault: () => {} });
      done();
    }, 700);
  });
  let wrapper2 = shallow(<SearchBar />);
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper2 = shallow(<SearchBar />);
  });
  test("should not submit the form", (done) => {
    const inputNoSubmitted = wrapper2.find("input");
    const value = "de";
    inputNoSubmitted.simulate("change", { target: { value } });
    setTimeout(() => {
      wrapper2.find("button").simulate("click", { preventDefault: () => {} });
      done();
    }, 600);
  });
});
