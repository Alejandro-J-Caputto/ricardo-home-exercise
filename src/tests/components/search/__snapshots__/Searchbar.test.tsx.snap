// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on searchBar component should render Searchbar correctly 1`] = `
<div
  className="searchBar"
>
  <form
    className="searchBar-group"
  >
    <input
      autoComplete="off"
      className="searchBar-group__input"
      id="search"
      name="search"
      onChange={[Function]}
      placeholder="Search Article"
      type="text"
    />
    <label
      className="searchBar-group__label"
      htmlFor="search"
    >
      Search Article
    </label>
    <button
      className="btn btn-search"
      disabled={true}
      onClick={[Function]}
    >
      <i
        className="fas fa-search"
      />
      <span>
        Search
      </span>
    </button>
  </form>
</div>
`;
